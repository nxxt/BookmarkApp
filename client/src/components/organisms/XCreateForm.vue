<template>
  <v-card
    :color="color"
    class="x-create-form"
    :style="{ width: `${isMobile ? '100%' : '90%'}` }"
  >
    <v-card-title>Create Bookmark</v-card-title>

    <v-card-text class="pt-6">
      <v-form ref="form" v-model="valid" lazy-validation>
        <h4>URL</h4>
        <v-text-field
          v-model="url"
          placeholder="Insert a link URL to any page."
          :rules="[isUrl]"
          validate-on-blur
        />
      </v-form>
    </v-card-text>

    <v-card-actions>
      <v-spacer />
      <x-tooltip-btn
        tooltip="save page"
        :disabled="url === '' || !valid"
        @click="onBtnClick"
      >
        save
      </x-tooltip-btn>
    </v-card-actions>

    <v-card-text class="d-flex justify-center">
      <!-- slot for progress circular -->
      <slot></slot>
    </v-card-text>
  </v-card>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator'
import XTooltipBtn from '@/components/atoms/XTooltipBtn.vue'

@Component({
  components: { XTooltipBtn }
})
export default class XCreateForm extends Vue {
  url: string = ''
  valid: boolean = false

  @Prop({ type: Boolean, default: false })
  isMobile!: boolean

  @Prop({ type: String, default: '' })
  color!: string

  onBtnClick() {
    const form = this.$refs.form as any
    if (form && form.validate()) {
      this.$emit('url-save-btn-click', this.url)
    }
  }

  isUrl(str: string) {
    if (str === '') return 'url is empty'
    const regex = new RegExp(
      "((https?|ftp)(://[-_.!~*'()a-zA-Z0-9;/?:@&=+$,%#]+))"
    )
    return regex.test(str) || 'invalid url'
  }
}
</script>
