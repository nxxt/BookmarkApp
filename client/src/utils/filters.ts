import IBookmark from '~~/types/IBookmark'
import ITag from '~~/types/ITag'
import IFolder from '~~/types/IFolder'

const filters: { [key: string]: Function } = {
  bookmarkTitle: (bookmark: IBookmark) => {
    return bookmark && bookmark.title ? bookmark.title : ''
  },
  tagName: (tag: ITag) => {
    return tag && tag.name ? `#${tag.name}` : ''
  },
  folderName: (folder: IFolder) => {
    return folder && folder.name ? folder.name : ''
  },
  textOverflow: (str: string) => {
    return str.length > 140 ? `${str.substring(0, 140)}â€¦` : str
  },
  dateFormat: (str: string) => {
    return str.substring(0, 10)
  }
}

export default filters
