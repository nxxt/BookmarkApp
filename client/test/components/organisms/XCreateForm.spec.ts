import Vuetify from 'vuetify'
import { createLocalVue, mount } from '@vue/test-utils'
import XCreateForm from '@/components/organisms/XCreateForm.vue'
import filters from '@/utils/filters'
import mockGetters from '@/utils/mock_getters'

const localVue = createLocalVue()
Object.keys(filters).forEach((key: string) =>
  localVue.filter(key, filters[key])
)
Object.keys(mockGetters).forEach((key: string) => {
  localVue.prototype[key] = mockGetters[key]
})

describe('XCreateForm.vue', () => {
  /****************
   * 準備
   ****************/
  // vuetify
  let vuetify: typeof Vuetify
  beforeEach(() => {
    vuetify = new Vuetify()
  })

  // propsで渡す値
  const propsData = {
    color: 'darker',
    isMobile: false
  }

  const factory = (options?: any) => {
    return mount(XCreateForm, {
      localVue,
      vuetify,
      ...options
    })
  }

  /****************
   * TEST
   ****************/
  it('props', () => {
    const wrapper = factory({ propsData })
    expect(wrapper.props()).toStrictEqual(propsData)
  })

  describe('template', () => {
    it('slot', () => {
      const wrapper = factory({
        slots: {
          default: '<div data-test="slotContent">slot content</div>'
        }
      })

      const slotContent = wrapper.find('[data-test="slotContent"]')
      expect(slotContent.exists()).toBe(true)
      expect(slotContent.text()).toBe('slot content')
    })

    it('text', () => {
      const wrapper = factory()
      const title = wrapper.find('.v-card__title')
      const label = wrapper.find('.v-card__text h4')
      const btn = wrapper.find('.v-btn__content')

      expect(title.text()).toBe('Create Bookmark')
      expect(label.text()).toBe('URL')
      expect(btn.text()).toBe('save')
    })

    it('isMobile', () => {
      const wrapper = factory()
      const elm = wrapper.find('.v-card')
      expect(elm.element.style.width).toBe('90%')
      wrapper.setProps({ isMobile: true })

      // propsセット後の変化を検査するにはnextTickが必要
      wrapper.vm.$nextTick(() => {
        expect(elm.element.style.width).toBe('100%')
      })
    })

    it('text-field', () => {
      const wrapper = factory()
      const input = wrapper.find('.v-text-field__slot input')
      const message = wrapper.find('.v-messages__wrapper')
      expect(input.attributes().placeholder).toBe(
        'Insert a link URL to any page.'
      )
      expect(message.text()).toBe('') // TODO:check
      expect(wrapper.vm.$data.url).toBe('')

      input.setValue('https://test.com')

      // propsセット後の変化を検査するにはnextTickが必要
      wrapper.vm.$nextTick(() => {
        expect(wrapper.vm.$data.url).toBe('https://test.com')
        expect(message.text()).toBe('')
      })
    })

    it('text-field empty', () => {
      const wrapper = factory()
      const field = wrapper.find('.v-text-field') // vmをもつ要素を取得するべし
      const fVm = field.vm as any
      expect(wrapper.vm.$data.url).toBe('')

      // field.setValue('')
      fVm.value = ''

      wrapper.vm.$nextTick(() => {
        fVm.$emit('change')

        expect(wrapper.vm.$data.url).toBe('')
        /* // labelがtransiton elementでtextがうまく取得できない
        const message = wrapper.find('.v-messages__wrapper')
        console.log(message.text())
        expect(message.text()).toBe('url is empty') */
      })
    })

    it('text-field invalid', () => {
      const wrapper = factory()
      const input = wrapper.find('.v-text-field__slot input')

      expect(wrapper.vm.$data.url).toBe('')

      input.setValue('fail_case')

      wrapper.vm.$nextTick(() => {
        expect(wrapper.vm.$data.url).toBe('fail_case')
        /* // labelがtransiton elementでtextがうまく取得できない
        const message = wrapper.find('.v-messages__message')
        expect(message.text()).toBe('invalid url') */
      })
    })

    it('btn clear disabled', () => {
      const wrapper = factory()
      const btn = wrapper.find('.v-btn')

      expect(btn.attributes().disabled).toBe('disabled')

      wrapper.setData({ url: 'https://test.com', valid: true })

      wrapper.vm.$nextTick(() => {
        expect(btn.attributes()).not.toContain('disabled')
      })
    })

    it('@click="onBtnClick"', () => {
      const mock = jest.fn()
      const wrapper = factory()
      wrapper.setMethods({ onBtnClick: mock })
      const btn = wrapper.find('.v-btn')
      // ↑【確認済】ここまでの時点でonBtnClickはmock化されている。

      // × btn.vm.$emit('click')  // triggerしてもmockのcallがカウントされない。
      // × btn.trigger('click')   // triggerしてもmockのcallがカウントされない。
      // set系のメソッド実行後はnextTickが必要？
      wrapper.vm.$nextTick(() => {
        // × btn.trigger('click') // triggerしてもmockのcallがカウントされない。WHY？
        btn.vm.$emit('click')
        expect(mock).toBeCalled()
      })
    })
  })

  describe('metohds', () => {
    it('onBtnClick', () => {
      const wrapper = factory({ propsData })
      const vm: any = wrapper.vm
      wrapper.setMethods({ isUrl: () => true })
      wrapper.setData({ url: 'test' })

      const event = jest.fn()
      vm.$on('url-save-btn-click', event)

      // 初期状態
      expect(event).toBeDefined() // イベントがundefinedでないこと
      expect(event).toHaveBeenCalledTimes(0) // eventが発生していないこと

      wrapper.vm.$nextTick(() => {
        // emit
        vm.onBtnClick()

        // イベント後
        expect(event).toHaveBeenCalledTimes(1) // イベントが親へ伝搬したこと
        const a = wrapper.emitted('url-save-btn-click')
        expect(a).toBeDefined()
        if (a) expect(a[0][0]).toEqual('test') // イベントのペイロードが正しいこと
      })
    })

    it('isUrl', () => {
      const wrapper = factory({ propsData })
      const vm: any = wrapper.vm
      expect(vm.isUrl('')).toBe('url is empty')
      expect(vm.isUrl('https://test.com')).toBe(true)
      expect(vm.isUrl('fail_test')).toBe('invalid url')
    })
  })

  describe('the entire template', () => {
    it('snapshot', () => {
      const wrapper = factory({ propsData })
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
})
