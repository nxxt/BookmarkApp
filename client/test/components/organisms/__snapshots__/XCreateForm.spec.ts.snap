// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`XCreateForm.vue the entire template snapshot 1`] = `
"<div class=\\"x-create-form v-card v-sheet theme--light darker\\" style=\\"width: 90%;\\">
  <div class=\\"v-card__title\\">Create Bookmark</div>
  <div class=\\"v-card__text pt-6\\">
    <form novalidate=\\"novalidate\\" class=\\"v-form\\">
      <h4>URL</h4>
      <div class=\\"v-input theme--light v-text-field v-text-field--placeholder\\">
        <div class=\\"v-input__control\\">
          <div class=\\"v-input__slot\\">
            <div class=\\"v-text-field__slot\\"><input id=\\"input-136\\" placeholder=\\"Insert a link URL to any page.\\" type=\\"text\\"></div>
          </div>
          <div class=\\"v-text-field__details\\">
            <div class=\\"v-messages theme--light\\">
              <div class=\\"v-messages__wrapper\\"></div>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>
  <div class=\\"v-card__actions\\">
    <div class=\\"spacer\\"></div> <span class=\\"v-tooltip v-tooltip--bottom\\"><!----><div><button type=\\"button\\" disabled=\\"disabled\\" class=\\"x-tooltip-btn v-btn v-btn--contained v-btn--disabled theme--light v-size--default\\"><span class=\\"v-btn__content\\">
      save
    </span></button>
  </div></span>
</div>
<div class=\\"v-card__text d-flex justify-center\\"></div>
</div>"
`;
